ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/util/DebugDataDumperTest.java

```diff
@@ -0,0 +1,232 @@
+  @Test
+  void saveInvalidBlockToFile_shouldSaveToFile(@TempDir Path tempDir) {
+    final DebugDataDumper manager = new DebugDataDumper(tempDir, true);
+    final SignedBeaconBlock block = dataStructureUtil.randomSignedBeaconBlock();
+    manager.saveInvalidBlockToFile(
+        block.getSlot(),
+        block.getRoot(),
+        block.sszSerialize(),
+        "reason",
+        Optional.of(new Throwable()));
+
+    final String fileName =
+        String.format("%s_%s.ssz", block.getSlot(), block.getRoot().toUnprefixedHexString());
+    final Path expectedFile = tempDir.resolve("invalid_blocks").resolve(fileName);
+    checkBytesSavedToFile(expectedFile, block.sszSerialize());
+  }
+
+  @Test
+  void saveInvalidBlockToFile_shouldNotSaveToFileWhenDisabled(@TempDir Path tempDir) {
+    final DebugDataDumper manager = new DebugDataDumper(tempDir, false);
+    final SignedBeaconBlock block = dataStructureUtil.randomSignedBeaconBlock();
+    manager.saveInvalidBlockToFile(
+        block.getSlot(), block.getRoot(), block.sszSerialize(), "reason", Optional.empty());
+    assertThat(manager.isEnabled()).isFalse();
+
+    final String fileName =
+        String.format("%s_%s.ssz", block.getSlot(), block.getRoot().toUnprefixedHexString());
+    final Path expectedFile = tempDir.resolve("invalid_blocks").resolve(fileName);
+    checkFileNotExist(expectedFile);
+  }
+
+  @Test
+  void saveBytesToFile_shouldNotThrowExceptionWhenNoDirectory(@TempDir Path tempDir) {
+    final DebugDataDumper manager = new DebugDataDumper(tempDir, true);
+    assertDoesNotThrow(
+        () -> {
+          final boolean success =
+              manager.saveBytesToFile(
+                  "object", Path.of("invalid").resolve("file.ssz"), Bytes.EMPTY);
+          assertThat(success).isTrue(); // creates directory
+        });
+  }
+
+  @Test
+  @DisabledOnOs(OS.WINDOWS) // Can't set permissions on Windows
+  void saveBytesToFile_shouldNotEscalateWhenIOException(@TempDir Path tempDir) {
+    final DebugDataDumper manager = new DebugDataDumper(tempDir, true);
+    final File invalidPath = tempDir.resolve("invalid").toFile();
+    assertThat(invalidPath.mkdirs()).isTrue();
+    assertThat(invalidPath.setWritable(false)).isTrue();
+    assertDoesNotThrow(
+        () -> {
+          final boolean success =
+              manager.saveBytesToFile(
+                  "object", Path.of("invalid").resolve("file.ssz"), Bytes.EMPTY);
+          assertThat(success).isFalse();
+        });
+  }
+
+  @Test
+  @DisabledOnOs(OS.WINDOWS) // Can't set permissions on Windows
+  void constructionOfDirectories_shouldDisableWhenFailedToCreate(@TempDir Path tempDir) {
+    assertThat(tempDir.toFile().setWritable(false)).isTrue();
+    final DebugDataDumper manager = new DebugDataDumper(tempDir, true);
+    assertThat(manager.isEnabled()).isFalse();
+  }
+
+  @Test
+  void formatOptionalTimestamp_shouldFormatTimestamp() {
+    final DebugDataDumper manager = new DebugDataDumper(Path.of("."), true);
```

Comment:
if we're going to do this we'll need to be using Tempdir, or we'll be polluting the working directory with folders constantly... i think similar elsewhere, this will be creating a bunch of folders.